File: TSOS/C#/lab_3/DSP_LW3/Matrix.cs
using System;

namespace DSP_LW3
{
    public static class Matrix
    {
        public static double[,] Sharpen3x3
        {
            get
            {
                return new double[,]
                {
                    { 0,  -1,  0, },
                    { -1,  5,  -1, },
                    { 0,  -1,  0, },
                };
            }
        }

        public static double[,] Sobel3x3Horizontal
        {
            get
            {
                return new double[,]
                {
                    { -1,  0,  1, },
                    { -2,  0,  2, },
                    { -1,  0,  1, },
                };
            }
        }

        public static double[,] Sobel3x3Vertical
        {
            get
            {
                return new double[,]
                {
                    {  1,  2,  1, },
                    {  0,  0,  0, },
                    { -1, -2, -1, },
                };
            }
        }

        public static (double[,], double) GetBoxBlurMatrix(int size)
        {
            double[,] matrix = new double[size, size];
            int offset = (size - 1) / 2;
            double sum = 0;
            for (int y = -offset; y <= offset; y++)
            {
                for (int x = -offset; x <= offset; x++)
                {
                    matrix[y + offset, x + offset] = 1;
                    sum += matrix[y + offset, x + offset];
                }
            }

            sum = 1 / sum;
            return (matrix, sum);
        }

        public static (double[,], double) GetMotionBlurMatrix(int size)
        {
            double[,] matrix = new double[size, size];
            int offset = (size - 1) / 2;
            double sum = 0;
            for (int y = -offset; y <= offset; y++)
            {
                for (int x = -offset; x <= offset; x++)
                {
                    if (x == y)
                    {
                        matrix[y + offset, x + offset] = 1;
                        sum += matrix[y + offset, x + offset];
                    }
                }
            }

            sum = 1 / sum;
            return (matrix, sum);
        }

        public static (double[,], double) GetGaussianMatrix(double sigma)
        {
            int size = (int)Math.Ceiling(6 * sigma);
            size = (size & 1) == 0 ? size + 1 : size;
            double[,] matrix = new double[size, size];
            int offset = (size - 1) / 2;
            double sum = 0;
            for (int y = -offset; y <= offset; y++)
            {
                for (int x = -offset; x <= offset; x++)
                {
                    matrix[y + offset, x + offset] = 1d / (2 * Math.PI * sigma * sigma) * Math.Exp(-((x * x) + (y * y)) / (2 * sigma * sigma));
                    sum += matrix[y + offset, x + offset];
                }
            }

            sum = 1 / sum;
            return (matrix, sum);
        }

        public static (double[], double) Get1DGaussianMatrix(double sigma)
        {
            int size = (int)Math.Ceiling(6 * sigma);
            size = (size & 1) == 0 ? size + 1 : size;
            double[] matrix = new double[size];
            int offset = (size - 1) / 2;
            double sum = 0;
            for (int x = -offset; x <= offset; x++)
            {
                matrix[x + offset] = 1d / (Math.Sqrt(2 * Math.PI) * sigma) * Math.Exp(-(x * x) / (2 * sigma * sigma));
                sum += matrix[x + offset];
            }

            sum = 1 / sum;
            return (matrix, sum);
        }
    }
}

File: TSOS/JS/lab_3/Matrix.js
// export const Sharpen3x3 = [
//   [0, -1, 0],
//   [-1, 5, -1],
//   [0, -1, 0],
// ];
export const Sobel3x3Horizontal = [
  [-1, 0, 1],
  [-2, 0, 2],
  [-1, 0, 1],
];
export const Sobel3x3Vertical = [
  [1, 2, 1],
  [0, 0, 0],
  [-1, -2, -1],
];

function createX_X_Matrix(size) {
  let result = new Array(size);
  result = result.fill([]).map(e => new Array(size).fill(undefined));
  return result;
}

export function GetBoxBlurMatrix(size) {
  const matrix = createX_X_Matrix(size);
  let sum = 0;
  const offset = Math.floor((size - 1) / 2);

  for (let y = -offset; y <= offset; y++) {
    for (let x = -offset; x <= offset; x++) {
      matrix[y + offset][x + offset] = 1;
      sum += matrix[y + offset][x + offset];
    }
  }
  sum = 1 / sum;

  return { matrix, sum };
}
// export function GetMotionBlurMatrix(size) {
//   const matrix = createX_X_Matrix(size);
//   let sum = 0;
//   const offset = Math.floor((size - 1) / 2);

//   for (let y = -offset; y <= offset; y++) {
//     for (let x = -offset; x <= offset; x++) {
//       if (x === y) {
//         matrix[y + offset][x + offset] = 1;
//         sum += matrix[y + offset][x + offset];
//       }
//     }
//   }
//   sum = 1 / sum;
//   return { matrix, sum };
// }
export function GetGaussianMatrix(sigma) {
  let size = Math.ceil(6 * sigma);
  size = (size & 1) === 0 ? size + 1 : size;
  const matrix = createX_X_Matrix(size);
  const offset = Math.floor((size - 1) / 2);

  let sum = 0;
  for (let y = -offset; y <= offset; y++) {
    for (let x = -offset; x <= offset; x++) {
      matrix[y + offset][x + offset] =
        (1 / (2 * Math.PI * sigma * sigma)) *
        Math.exp(-(x * x + y * y) / (2 * sigma * sigma));
      sum += matrix[y + offset][x + offset];
    }
  }
  sum = 1 / sum;
  return { matrix, sum };
}
